src/arch/riscv/pagetable_walker.cc

```diff
@@ -189,10 +197,44 @@ Walker::WalkerState::initState(ThreadContext * _tc,
     mode = _mode;
     timing = _isTiming;
     // fetch these now in case they change during the walk
+    memaccess = functional ?
+        walker->tlb->getMemAccessInfo(tc, mode, (Request::ArchFlagsType)0):
+        walker->tlb->getMemAccessInfo(tc, mode, req->getArchFlags());
+    pmode = memaccess.priv;
     status = tc->readMiscReg(MISCREG_STATUS);
-    pmode = walker->tlb->getMemPriv(tc, mode);
-    satp = tc->readMiscReg(MISCREG_SATP);
-    assert(satp.mode == AddrXlateMode::SV39);
+    MISA misa = tc->readMiscReg(MISCREG_ISA);
+
+    // Find SATP
+    // If no rvh or effective V = 0, base is SATP
+    // otherwise base is VSATP (effective V=1)
+    satp = (!misa.rvh || !memaccess.virt) ?
+            tc->readMiscReg(MISCREG_SATP) :
+            tc->readMiscReg(MISCREG_VSATP);
+
+    // If effective V = 1, also read HGATP for
+    // G-stage because we will perform a two-stage translation.
+    hgatp = (misa.rvh && memaccess.virt) ?
+            tc->readMiscReg(MISCREG_HGATP) :
+            (RegVal)0;
+
+    // TODO move this somewhere else
+    // VSATP mode might be bare, but we still
+    // will have to go through G-stage
+    // assert(satp.mode == AddrXlateMode::SV39);
+
+    // If functional entry.vaddr will be set
+    // in start functional (req == NULL)
+    entry.vaddr = functional ?
+        (Addr)0 :
+        req->getVaddr();
+
+    entry.asid = satp.asid;
+    if (debug_condition()) {
+        DPRINTF(PageTableWalker, "======================================\n");
```

Comment:
Is this absolutely necessary? I think prints like this end up spamming the output.